JumpscaleLib/tools/blockchain/tfchain/encoding/binary.py

```diff
@@ -0,0 +1,122 @@
+"""
+Module to support binary encoding for tfchain
+"""
+
+
+SLICE_LENGTH_1BYTES_UPPERLIMIT = pow(2, 7)
+SLICE_LENGTH_2BYTES_UPPERLIMIT = pow(2, 14)
+SLICE_LENGTH_3BYTES_UPPERLIMIT = pow(2, 21)
+SLICE_LENGTH_4BYTES_UPPERLIMIT = pow(2, 29)
+
+UINT_1BYTE_UPPERLIMIT = pow(2, 8) - 1
+UINT_2BYTE_UPPERLIMIT = pow(2, 16) - 1
+UINT_3BYTE_UPPERLIMIT = pow(2, 24) - 1
+UINT_4BYTE_UPPERLIMIT = pow(2, 32) - 1
+
+
+class BinaryEncoder:
+    """
+    Support binary encoding for tfchain
+    """
+    @staticmethod
+    def encode(value, type_=None):
+        """
+        Enocdes a value to its binary representation
+
+        @param value: Value to be encoded
+        @param type_: Type of the value, if not given, the type will be detected automatically
+        """
+        result = bytearray()
+        if value is None:
+            return result
+        if type_ == 'slice':
+            # slice is a variable size array, which means we need to prefix the size of the array and then we can encode the list itself
+            result = SliceBinaryEncoder.encode(value)
+        elif type_ == 'currency':
+            nbytes, rem = divmod(value.bit_length(), 8)
+            if rem:
+                nbytes += 1
+            result.extend(encode(nbytes))
+            result.extend(value.to_bytes(nbytes, byteorder='big'))
+        elif type_ == 'hex':
+            result.extend(bytearray.fromhex(value))
+        elif type_ is None:
+            # try to figure out the type of the value
+            value_type = type(value)
+            if value_type in (bytes, bytearray):
+                result.extend(value)
+            elif value_type is int:
+                result = IntegerBinaryEncoder.encode(value)
+            elif value_type is bool:
+                result = bytearray([1]) if value else bytearray([0])
+            elif value_type in (list, set, tuple, frozenset):
+                for item in value:
+                    result.extend(encode(item))
+            else:
+                if hasattr(value, 'binary'):
+                    result.extend(value.binary)
+        else:
+            raise ValueError('Cannot binary encode value with unknown type')
+        return result
+
+
+
+class SliceBinaryEncoder:
+    """
+    Support binary encoding for slice type
+    Slice is an array with no pre-defined size
+    """
+
+    @staticmethod
+    def encode_length(length):
+        """
+        Encodes the length of the slice
+        """
+        if length < SLICE_LENGTH_1BYTES_UPPERLIMIT:
+            return IntegerBinaryEncoder.encode(length << 1)
+        elif length < SLICE_LENGTH_2BYTES_UPPERLIMIT:
+            return IntegerBinaryEncoder.encode((1 | length << 2))
```

Comment:
Not a Python programmer, but aren't these double braces useless?
```suggestion
            return IntegerBinaryEncoder.encode(1 | length << 2)
```

Reply:
Yes it is useless. It's a leftover from a version of the code where the encode function used to accept another parameter and the extra braces were added to make it more readable. Thanks for the feedback